npm/test/folderinforetrieval.test.ts

```diff
@@ -57,60 +80,299 @@ const mockFormHTML = `
+      "test",
+      mockGoToNextPage
+    )
+  }
+  selectFolderViaButton: () => void = () => {
+    triggerInputEvent(this.folderRetriever!, "change")
+  }
+}
 
 const mockGoToNextPage = jest.fn()
 
-const triggerInputEvent: (
-  element: HTMLInputElement,
+const triggerInputEvent: (element: HTMLElement, domEvent: string) => boolean = (
+  element: HTMLElement,
   domEvent: string
-) => void = (element: HTMLInputElement, domEvent: string) => {
+) => {
   const event = new CustomEvent(domEvent)
-  element.dispatchEvent(event)
+  return element.dispatchEvent(event)
 }
 
-test("folder retriever updates the page with correct folder information if there are 1 or more files", () => {
-  document.body.innerHTML = mockFormHTML
+const dummyFolder = {
+  lastModified: 2147483647,
+  name: "Mock Folder",
+  size: 0,
+  type: ""
+} as File
 
-  const dummyFile = {
-    webkitRelativePath: "Parent_Folder/testfile"
-  } as TdrFile
+const dummyFile = {
+  lastModified: 2147483647,
+  name: "Mock File",
+  size: 3008,
+  type: "pdf"
+} as File
 
-  const uploadForm: HTMLFormElement | null = document.querySelector(
-    "#file-upload-form"
+const dummyIFileWithPath = {
+  file: dummyFile,
+  path: "Parent_Folder",
+  webkitRelativePath: "Parent_Folder"
+} as IFileWithPath
+
+test("Input button updates the page with correct folder information if there are 1 or more files in folder", () => {
+  const mockDom = new MockDom()
+  mockDom.fileUploader.initialiseFormListeners()
+  mockDom.uploadForm!.files = { files: [dummyIFileWithPath] }
+  mockDom.selectFolderViaButton()
+
+  expect(mockDom.folderRetrievalSuccessMessage!).not.toHaveClass("hide")
+  expect(mockDom.folderRetrievalfailureMessage!).toHaveClass("hide")
+  expect(mockDom.folderNameElement!.textContent).toStrictEqual("Parent_Folder")
+  expect(mockDom.folderSizeElement!.textContent).toStrictEqual("1 file")
+})
+
+test("drop event is triggerable", () => {
+  const mockDom = new MockDom()
+  expect(triggerInputEvent(mockDom.dropzone!, "drop")).toBe(true)
+})
+
+test("dropzone updates the page with correct folder information if there are 1 or more files in folder", async () => {
+  const mockDom = new MockDom()
+  const dragEventClass = mockDom.addFilesToDragEvent(
+    [dummyFolder],
+    mockDom.dataTransferItem
   )
-  const folderRetriever: HTMLInputElement | null = document.querySelector(
-    "#file-selection"
+  const dragEvent = new dragEventClass()
+  await mockDom.form.handleDropppedItems(dragEvent)
+
+  expect(mockDom.folderRetrievalSuccessMessage!).not.toHaveClass("hide")
+  expect(mockDom.folderRetrievalfailureMessage!).toHaveClass("hide")
+  expect(mockDom.folderNameElement!.textContent).toStrictEqual("Mock Folder")
+  expect(mockDom.folderSizeElement!.textContent).toStrictEqual("2 files")
+})
+
+test("dropzone updates the page with an error if there are no files in folder", async () => {
+  const mockDom = new MockDom()
+  mockDom.entries.pop() // remove entries from previous test
```

Comment:
Is it possible to reset MockDom with each test, or make it not have mutable state? This line implies that you can only run the tests in a specific order. It's much easier to update tests that are completely independent.

Reply:
It's not so much that it's not resetting, it's that the mocked files are hardcoded into the MockDom class, so to simulate zero files, they have to be removed for this test. I've moved it out to the class constructor and passed in a numberOfFiles argument. It's not really removing the logic but it's keeping it within the class.

I agree with your point on the tests but the functions for these File API methods is a mess of callbacks, symbol arrays and all sorts of fun things so as you say, they're ok for now and we can revisit them later.

Reply:
Thanks @MancunianSam! Yes, I added the files directly to the mockDom to save having to add them each time; my comment could have been a little clearer though, I think they did share state initially, before I put them inside the Class

Reply:
Parameterising it is good, thanks. 👍